snyk/broker
# feat: add ability to create a credential array

- [✅] Ready for review
- [✅] Follows [CONTRIBUTING](https://github.com/snyk/broker/blob/master/.github/CONTRIBUTING.md) rules

#### What does this PR do?
Users can now specify a credential pool by adding "_POOL" to the
environment variable name, and providing a comma-separated list of
credentials as the value, which will get turned into an array
internally. For example, to provide multiple GitHub tokens, you would
do:

GITHUB_TOKEN_POOL=token1, token2, token3, ...

This will then create an array, and anywhere that GITHUB_TOKEN is
referenced (e.g., in accept.json) it will instead pick a token from that
list (picking the first then second then third then back to the
beginning, etc).

It will also create an array of any other variables that reference it,
like BROKER_CLIENT_VALIDATION_AUTHORIZATION_HEADER - it will create a
BROKER_CLIENT_VALIDATION_AUTHORIZATION_HEADER_POOL value, which is
used by the systemcheck endpoint.

This includes a breaking change to the /systemcheck endpoint as it now
returns a JSON array rather than an object - this is to accommodate
checking multiple credentials at once.

#### Where should the reviewer start?
Start in `config.js`, which has to parse the pools, then `replace-vars.js`, which handles the actual replacing. Then `index.js` for the `/systemcheck` endpoint and `filter.js` has some minor changes applied.


#### How should this be manually tested?
Start the Broker Client with an `_POOL` environment variable for whichever provider you're using. Verify the `/systemcheck` endpoint picks up both credentials. Verify multiple requests work. Add one good and one bad credential, and verify 50% of requests fail.

#### Any background context you want to provide?


#### Screenshots


#### Additional questions
